### PR TITLE
Release v3.1.0

### PR DESCRIPTION
**Pull request by hollasch**

### CODE REVIEW

File: books/RayTracingTheRestOfYourLife.html [hunk 1]
```diff
@@ -1679,19 +1679,7 @@
...
-                : x0(_x0), x1(_x1), z0(_z0), z1(_z1), k(_k), mp(mat) {};
-
-            virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
-
-            virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
-                output_box = aabb(point3(x0,k-0.0001,z0), point3(x1, k+0.0001, z1));
-                return true;
-            }
-
+            ...
```
Comment (dafhi) [inline books/RayTracingTheRestOfYourLife.html:1682]: yeah baby!

File: books/RayTracingTheNextWeek.html [hunk 2]
```diff
@@ -2136,7 +2145,9 @@
             virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
             virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
-                output_box =  aabb(point3(k-0.0001, y0, z0), point3(k+0.0001, y1, z1));
+                // The bounding box must have non-zero width in each dimension, so pad the X
```
Comment (dafhi) [inline books/RayTracingTheNextWeek.html:2148]: nice

File: books/RayTracingTheNextWeek.html [hunk 3]
```diff
@@ -2013,7 +2018,9 @@
             virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
             virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
-                output_box =  aabb(point3(x0,y0, k-0.0001), point3(x1, y1, k+0.0001));
+                // The bounding box must have non-zero width in each dimension, so pad the Z
```
Comment (dafhi) [inline books/RayTracingTheNextWeek.html:2021]: *thumbs*

File: books/RayTracingTheNextWeek.html [hunk 4]
```diff
@@ -2117,7 +2124,9 @@
             virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
             virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
-                output_box =  aabb(point3(x0,k-0.0001,z0), point3(x1, k+0.0001, z1));
+                // The bounding box must have non-zero width in each dimension, so pad the Y
```
Comment (dafhi) [inline books/RayTracingTheNextWeek.html:2127]: yep

File: books/RayTracingTheNextWeek.html [hunk 5]
```diff
@@ -1999,6 +1999,11 @@
 
 It is a hit if $x_0 < x < x_1$ and $y_0 < y < y_1$.
 
+Because our rectangles are axis-aligned, their bounding boxes will have an infinitely-thin side.
```
Comment (dafhi) [inline books/RayTracingTheNextWeek.html:2002]: good

File: books/RayTracingTheNextWeek.html [hunk 6]
```diff
@@ -7,7 +7,7 @@
                                            Peter Shirley
                           edited by Steve Hollasch and Trevor David Black
                                                 <br>
-                                     Version 3.1.0-wip, 2020-XX-XX
+                                     Version 3.1.0, 2020-05-03
```
Comment (dafhi) [inline books/RayTracingTheNextWeek.html:10]: thumbs

File: books/RayTracingInOneWeekend.html [hunk 7]
```diff
@@ -7,7 +7,7 @@
                                            Peter Shirley
                           edited by Steve Hollasch and Trevor David Black
                                                 <br>
-                                     Version 3.1.0-wip, 2020-XX-XX
+                                     Version 3.1.0, 2020-05-03
```
Comment (dafhi) [inline books/RayTracingInOneWeekend.html:10]: thumbs

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,11 +2,12 @@ Change Log -- Ray Tracing in One Weekend
...
 
 ----------------------------------------------------------------------------------------------------
-# v3.1.0 (in progress)
+# v3.1.0 (2020-05-03)
 
 ### Common
 - Fix: Scattered improvements to the text.
-- New: subchapters throughout all three books (#267)
+- New: Subchapters throughout all three books (#267)
+- New: Add explanation for padding `aarect` in the zero dimension (#488)
```
Comment (dafhi) [inline CHANGELOG.md:10]: i confirm

File: CHANGELOG.md [hunk 9]
```diff
@@ -2,11 +2,12 @@ Change Log -- Ray Tracing in One Weekend
 ====================================================================================================
 
 ----------------------------------------------------------------------------------------------------
-# v3.1.0 (in progress)
+# v3.1.0 (2020-05-03)
```
Comment (dafhi) [inline CHANGELOG.md:5]: thumb

File: books/RayTracingTheRestOfYourLife.html [hunk 10]
```diff
@@ -7,7 +7,7 @@
                                            Peter Shirley
                           edited by Steve Hollasch and Trevor David Black
                                                 <br>
-                                     Version 3.1.0-wip, 2020-XX-XX
+                                     Version 3.1.0, 2020-05-03
```
Comment (dafhi) [inline books/RayTracingTheRestOfYourLife.html:10]: thumb

File: src/TheNextWeek/aarect.h [hunk 11]
```diff
@@ -27,6 +27,8 @@ class xy_rect: public hittable {
         virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
         virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
+            // The bounding box must have non-zero width in each dimension, so pad the Z
```
Comment (dafhi) [inline src/TheNextWeek/aarect.h:30]: yep

File: src/TheNextWeek/aarect.h [hunk 12]
```diff
@@ -47,6 +49,8 @@ class xz_rect: public hittable {
         virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
         virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
+            // The bounding box must have non-zero width in each dimension, so pad the Y
```
Comment (dafhi) [inline src/TheNextWeek/aarect.h:52]: yep

File: src/TheNextWeek/aarect.h [hunk 13]
```diff
@@ -67,6 +71,8 @@ class yz_rect: public hittable {
         virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
         virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
+            // The bounding box must have non-zero width in each dimension, so pad the X
```
Comment (dafhi) [inline src/TheNextWeek/aarect.h:74]: yep

File: src/TheRestOfYourLife/aarect.h [hunk 14]
```diff
@@ -27,6 +27,8 @@ class xy_rect: public hittable {
         virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
         virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
+            // The bounding box must have non-zero width in each dimension, so pad the Z
```
Comment (dafhi) [inline src/TheRestOfYourLife/aarect.h:30]: yep

File: src/TheRestOfYourLife/aarect.h [hunk 15]
```diff
@@ -47,6 +49,8 @@ class xz_rect: public hittable {
         virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
         virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
+            // The bounding box must have non-zero width in each dimension, so pad the Y
```
Comment (dafhi) [inline src/TheRestOfYourLife/aarect.h:52]: yep

File: src/TheRestOfYourLife/aarect.h [hunk 16]
```diff
@@ -84,6 +88,8 @@ class yz_rect: public hittable {
         virtual bool hit(const ray& r, double t0, double t1, hit_record& rec) const;
 
         virtual bool bounding_box(double t0, double t1, aabb& output_box) const {
+            // The bounding box must have non-zero width in each dimension, so pad the X
```
Comment (dafhi) [inline src/TheRestOfYourLife/aarect.h:91]: yep

File: books/RayTracingTheRestOfYourLife.html [hunk 17]
```diff
@@ -1710,12 +1698,6 @@
                 auto random_point = point3(random_double(x0,x1), k, random_double(z0,z1));
                 return random_point - origin;
             }
-    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
```
Comment (dafhi) [inline books/RayTracingTheRestOfYourLife.html:1713]: i r confused here but seems like u know what to do here

**Comment by hollasch** [comment]
Looks like I accidentally branched this one off of the aarect fix. Oh well. Ignore the first commit; that's covered in PR #511.

**Comment by trevordblack** [comment]
CHANGELOG needs a date for v3.0.3 (did we have one?), or subsume into 3.1.0

**Comment by hollasch** [comment]
Merged into dev-minor

**Pull request closed without merging**